File: joycv/cv/slice_image.py
import os
import glob
import math
import time
from datetime import datetime

import cv2
import numpy as np

def slice_single(img,mask,bbox,target_size):
    x0 = int(max(math.floor(bbox[0]) - 1, 0))
    y0 = int(max(math.floor(bbox[1]) - 1, 0))
    mask_img = img[y0:y0 + mask.shape[0], x0:x0 + mask.shape[1]]

    masked_image = cv2.bitwise_and(mask_img, mask_img, mask=mask)
    #masked_image=mask_img
    # Replace black area with the specified background color
    background_color = np.full(mask_img.shape, [113, 119, 52], dtype=np.uint8)
    background_mask = cv2.bitwise_and(background_color, background_color, mask=cv2.bitwise_not(mask))
    masked_image = cv2.add(masked_image, background_mask)

    # Scale down if necessary
    if max(masked_image.shape) > target_size:
        #print(f"Scale down image {img_name}_{index} to target_size")
        scale_factor = target_size / max(masked_image.shape)
        new_size = (int(masked_image.shape[1] * scale_factor), int(masked_image.shape[0] * scale_factor))
        masked_image = cv2.resize(masked_image, new_size)

    # Place the masked image on a 224x224 canvas with the background color
    canvas = np.full((target_size, target_size, 3), [113, 119, 52], dtype=np.uint8)
    x_offset = (target_size - masked_image.shape[1]) // 2
    y_offset = (target_size - masked_image.shape[0]) // 2
    canvas[y_offset:y_offset + masked_image.shape[0], x_offset:x_offset + masked_image.shape[1]] = masked_image
    return canvas,(x_offset,y_offset)

def slice_image(img,bboxes, labels, masks, img_path, output_path):
    
    start_time = time.time()
    
    inference_time = time.time() - start_time
    target_size=224
    img_name = os.path.splitext(os.path.basename(img_path))[0]
    indices = [i for i in range(len(bboxes))]
    
    for index, bbox, label_id in zip(indices, bboxes, labels):
        [left, top, right, bottom], score = bbox[0:4].astype(int), bbox[4]
        if score < 0.4:
            continue

        if masks[index].size:
            mask = masks[index]
            sub_img,offset_xy = slice_single(img,mask,bbox,target_size)
            save_path=output_path#os.path.join(output_path,"slice")
            if not os.path.exists(save_path):
                os.makedirs(save_path)
            save_path = os.path.join(save_path, f"{img_name}{score}_{index}.png")
            print(f"slicing {save_path}")
            cv2.imwrite(save_path, sub_img)



    return inference_time

global_counter=0
import math

def save_single_slice(output_path,sub_img,img_name,score,index):
    global global_counter
    
    sub_path=math.floor(global_counter/3000)
    save_path=os.path.join(output_path,f"group_{sub_path}")
    if not os.path.exists(save_path):
        os.makedirs(save_path)
    save_path = os.path.join(save_path, f"{img_name}{score}_{index}.png")
    print(f"saving sliced: {save_path}")
    cv2.imwrite(save_path, sub_img)
    global_counter=global_counter+1

File: joycv/cv/slice_image.cpp
#include <opencv2/opencv.hpp>
#include <vector>
#include <filesystem> // C++17
#include <chrono>
#include <cmath>

cv::Mat slice_single(const cv::Mat &img, const cv::Mat &mask, const std::vector<float> &bbox, int target_size) {
    int x0 = std::max(static_cast<int>(std::floor(bbox[0]) - 1), 0);
    int y0 = std::max(static_cast<int>(std::floor(bbox[1]) - 1), 0);
    cv::Rect roi(x0, y0, mask.cols, mask.rows);
    cv::Mat mask_img = img(roi);

    cv::Mat masked_image = cv::Mat::zeros(mask_img.size(), mask_img.type());
    mask_img.copyTo(masked_image, mask);

    cv::Mat background_color = cv::Mat::ones(mask_img.size(), mask_img.type()) * cv::Vec3b(113, 119, 52);
    cv::Mat background_mask;
    cv::bitwise_not(mask, background_mask);
    background_color.copyTo(background_mask, background_mask);

    cv::add(masked_image, background_mask, masked_image);

    if (std::max(masked_image.rows, masked_image.cols) > target_size) {
        float scale_factor = static_cast<float>(target_size) / std::max(masked_image.rows, masked_image.cols);
        cv::Size new_size(masked_image.cols * scale_factor, masked_image.rows * scale_factor);
        cv::resize(masked_image, masked_image, new_size);
    }

    cv::Mat canvas = cv::Mat::ones(target_size, target_size, CV_8UC3) * cv::Vec3b(113, 119, 52);
    int x_offset = (target_size - masked_image.cols) / 2;
    int y_offset = (target_size - masked_image.rows) / 2;
    cv::Rect roi_to(x_offset, y_offset, masked_image.cols, masked_image.rows);
    masked_image.copyTo(canvas(roi_to));

    return canvas;
}

void slice_image(const cv::Mat &img, const std::vector<std::vector<float>> &bboxes, 
                 const std::vector<int> &labels, const std::vector<cv::Mat> &masks,
                 const std::string &img_path, const std::string &output_path) {
    int target_size = 256;
    auto start_time = std::chrono::high_resolution_clock::now();

    std::filesystem::path path(img_path);
    std::string img_name = path.stem();

    for (int i = 0; i < bboxes.size(); i++) {
        std::vector<float> bbox = bboxes[i];
        if (bbox[4] < 0.4)
            continue;

        if (!masks[i].empty()) {
            cv::Mat sub_img = slice_single(img, masks[i], bbox, target_size);
            std::filesystem::path save_path = output_path;
            save_path /= img_name + std::to_string(bbox[4]) + "_" + std::to_string(i) + ".png";
            
            if (!std::filesystem::exists(save_path)) {
                std::filesystem::create_directories(save_path);
            }

            cv::imwrite(save_path.string(), sub_img);
        }
    }

    auto end_time = std::chrono::high_resolution_clock::now();
    auto inference_time = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time).count();

    std::cout << "Inference time: " << inference_time << "ms" << std::endl;
}
